<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />

  <title>The inferCNV method &mdash; infercnvpy  documentation</title>

      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/scanpy.css" type="text/css" />
      <link rel="stylesheet" href="_static/typehints.css" type="text/css" />

  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>


    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API" href="api.html" />
    <link rel="prev" title="Reproduce the heatmap from inferCNV" href="tutorials/reproduce_infercnv.html" />
 


</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> infercnvpy
          </a>
              <div class="version">
                0.2.1.dev13
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">The inferCNV method</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#computation-steps">Computation steps</a></li>
<li class="toctree-l2"><a class="reference internal" href="#preparing-input-data">Preparing input data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="zzz_bibliography.html">Bibliography</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">infercnvpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>The inferCNV method</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com/icbi-lab/infercnvpy/blob/master/docs/infercnv.rst" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="the-infercnv-method">
<span id="infercnv-method"></span><h1>The inferCNV method<a class="headerlink" href="#the-infercnv-method" title="Permalink to this headline"></a></h1>
<p>Essentially, this package is a Python reimplementation of
<a class="reference external" href="https://github.com/broadinstitute/inferCNV/">infercnv</a>. It mostly follows the computation steps
outlined <a class="reference external" href="https://github.com/broadinstitute/inferCNV/wiki/Running-InferCNV">here</a>,
with minor modifications. The computation steps are outlined below.
By making use of <code class="docutils literal notranslate"><span class="pre">numpy</span></code>, <code class="docutils literal notranslate"><span class="pre">scipy</span></code>, and sparse matrices,
it is a lot more computationally efficient.</p>
<section id="computation-steps">
<h2>Computation steps<a class="headerlink" href="#computation-steps" title="Permalink to this headline"></a></h2>
<p>The function parameters are documented at <a class="reference internal" href="generated/infercnvpy.tl.infercnv.html#infercnvpy.tl.infercnv" title="infercnvpy.tl.infercnv"><code class="xref py py-func docutils literal notranslate"><span class="pre">infercnvpy.tl.infercnv()</span></code></a>.</p>
<ol class="arabic">
<li><p>Subtract the reference gene expression from all cells. Since the data is in log
space, this effectively computes the log fold change. If references for
multiple categories are available (i.e. multiple values are specified to
<code class="docutils literal notranslate"><span class="pre">reference_cat</span></code>), the log fold change is “bounded”:</p>
<blockquote>
<div><ul class="simple">
<li><p>Compute the mean gene expression for each category separately.</p></li>
<li><p>Values that are within the minimum and the maximum of the mean of all
references, receive a log fold change of 0, since they are not considered
different from the background.</p></li>
<li><p>From values smaller than the minimum of the mean of all references, subtract that minimum.</p></li>
<li><p>From values larger than the maximum of the mean of all references, subtract that maximum.</p></li>
</ul>
</div></blockquote>
<p>This procedure avoids calling false positive CNV regions due to cell-type specific
expression of clustered gene regions (e.g. Immunoglobulin- or HLA genes in different
immune cell types).</p>
</li>
<li><p>Clip the fold changes at <code class="docutils literal notranslate"><span class="pre">-lfc_cap</span></code> and <code class="docutils literal notranslate"><span class="pre">+lfc_cap</span></code>.</p></li>
<li><p>Smooth the gene expression by genomic position. Computes the average over a
running window of length <code class="docutils literal notranslate"><span class="pre">window_size</span></code>. Compute only every nth window
to save time &amp; space, where n = <code class="docutils literal notranslate"><span class="pre">step</span></code>.</p></li>
<li><p>Center the smoothed gene expression by cell, by subtracting the median of each cell
from each cell.</p></li>
<li><p>Perform noise filtering. Values <code class="docutils literal notranslate"><span class="pre">&lt;</span> <span class="pre">dynamic_theshold</span> <span class="pre">*</span> <span class="pre">STDDEV</span></code> are set to 0,
where <code class="docutils literal notranslate"><span class="pre">STDDEV</span></code> is the standard deviation of the smoothed gene expression</p></li>
<li><p>Smooth the final result using a median filter.</p></li>
</ol>
</section>
<section id="preparing-input-data">
<span id="input-data"></span><h2>Preparing input data<a class="headerlink" href="#preparing-input-data" title="Permalink to this headline"></a></h2>
<p>The <a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.7.dev1+g33006dc)"><code class="xref py py-class docutils literal notranslate"><span class="pre">anndata.AnnData</span></code></a> object should already be filtered for low-quality cells.
<code class="docutils literal notranslate"><span class="pre">adata.X</span></code> needs to be normalized and log-transformed. The method should be
fairly robust to different normalization methods (<a class="reference external" href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.normalize_total.html#scanpy.pp.normalize_total" title="(in Scanpy v1.8.2)"><code class="xref py py-func docutils literal notranslate"><span class="pre">scanpy.pp.normalize_total()</span></code></a>, scran, etc.).</p>
<p>The method requires a “reference” value to which the expression of genomic
regions is compared. If your dataset contains different cell types and includes
both tumor and normal cells, the average of all cells can be used as reference.
This is the default.</p>
<p>If you already know which cells are “normal”, you can provide a column
from <code class="docutils literal notranslate"><span class="pre">adata.obs</span></code> to <code class="docutils literal notranslate"><span class="pre">reference_key</span></code> that contains the annotation. <code class="docutils literal notranslate"><span class="pre">reference_cat</span></code>
specifies one or multiple values in <code class="docutils literal notranslate"><span class="pre">reference_key</span></code> that refer to normal cells.</p>
<p>Alternatively, you can specify a numpy array with average gene expression values
(for instance, derived from a different dataset), to <code class="docutils literal notranslate"><span class="pre">reference</span></code> which will be used
as reference instead. This array needs to align to the <code class="docutils literal notranslate"><span class="pre">var</span></code> axis of adata.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="tutorials/reproduce_infercnv.html" class="btn btn-neutral float-left" title="Reproduce the heatmap from inferCNV" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="api.html" class="btn btn-neutral float-right" title="API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Gregor Sturm..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>